sway-lsp/src/core/session.rs

```diff
@@ -171,25 +170,24 @@ impl Session {
 
             let ast_res = CompileResult::new(typed, warnings, errors);
             let typed_program = self.compile_res_to_typed_program(&ast_res)?;
+            let ctx = ParseContext::new(&self.token_map, engines);
 
             // The final element in the results is the main program.
             if i == results_len - 1 {
                 // First, populate our token_map with sway keywords.
-                let lexed_tree = LexedTree::new(&self.token_map);
-                lexed_tree.parse(&lexed);
+                lexed_tree::parse(&lexed, &ctx);
 
                 // Next, populate our token_map with un-typed yet parsed ast nodes.
-                let parsed_tree = ParsedTree::new(type_engine, &self.token_map);
+                let parsed_tree = ParsedTree::new(&ctx);
                 parsed_tree.collect_module_spans(&parsed);
-                self.parse_ast_to_tokens(&parsed, |an| parsed_tree.traverse_node(an));
+                self.parse_ast_to_tokens(&parsed, &ctx, |an, _ctx| parsed_tree.traverse_node(an));
 
                 // Finally, create runnables and populate our token_map with typed ast nodes.
                 self.create_runnables(typed_program);
 
-                let typed_tree =
-                    TypedTree::new(engines, &self.token_map, &typed_program.root.namespace);
+                let typed_tree = TypedTree::new(&ctx, &typed_program.root.namespace);
                 typed_tree.collect_module_spans(typed_program);
-                self.parse_ast_to_typed_tokens(typed_program, |node| {
+                self.parse_ast_to_typed_tokens(typed_program, &ctx, |node, _ctx| {
```

Comment:
It might be a little cleaner to move `parse_ast_to_typed_tokens` into `TypedTree` and have functions that expose the parsing. Something like:

```
impl<'a> TypedTree<'a> {
    ...
    /// Parse the [ty::TyProgram] AST to populate the [TokenMap] with typed AST nodes.
    fn parse_ast_to_typed_tokens(&self, typed_program: &ty::TyProgram, f: impl Fn(&ty::TyAstNode)) {
        let root_nodes = typed_program.root.all_nodes.iter();
        let sub_nodes = typed_program
            .root
            .submodules
            .iter()
            .flat_map(|(_, submodule)| submodule.module.all_nodes.iter());

        root_nodes.chain(sub_nodes).for_each(|n| f(n));
    }

    pub fn parse(&self, typed_program: &TyProgram, is_main_program: bool) {
        if is_main_program {
            self.collect_module_spans(typed_program);
            self.parse_ast_to_typed_tokens(typed_program, |node| self.traverse_node(node));
        } else {
            self.parse_ast_to_typed_tokens(typed_program, |node| {
                dependency::collect_typed_declaration(node, self.ctx)
            });
        }
    }
    ...
}
```

A similar thing could be done for `ParsedTree`.

Then this would read:
```
            let parsed_tree = ParsedTree::new(&ctx);
            let typed_tree = TypedTree::new(&ctx, &typed_program.root.namespace);

            // The final element in the results is the main program.
            let is_main_program = i == results_len - 1;

            parsed_tree.parse(parsed, is_main_program);
            typed_tree.parse(typed_program, is_main_program);

            // The final element in the results is the main program.
            if is_main_program {
                // First, populate our token_map with sway keywords.
                lexed_tree::parse(&lexed, &ctx);

                // Next, populate the runnables map.
                self.create_runnables(typed_program);

                // Finally, save the lexed, parsed and typed trees.
                self.save_lexed_program(lexed.to_owned().clone());
                self.save_parsed_program(parsed.to_owned().clone());
                self.save_typed_program(typed_program.to_owned().clone());

                diagnostics = get_diagnostics(&ast_res.warnings, &ast_res.errors);
            }
```

Reply:
this is a good idea, i'll look at adding something along these lines in the follow up PR which refactors typed_program. :)